### PR TITLE
use System.CommandLine to parse command line options

### PR DESCRIPTION
**Pull request by 928PJY**

### CODE REVIEW

**Comment by superyyrrzz** [comment]
change all `ops.yml`/`docfx.yaml` to `docfx.yml`

File: src/docfx/cli/Program.cs [hunk 1]
```diff
@@ -9,23 +10,43 @@ internal class Program
...
+                // Restore command
+                syntax.DefineCommand("restore", ref command, "restores dependencies before build");
+                syntax.DefineParameter("source", ref source, "docset path that contains docfx.yml");
+
+                // Build command
+                syntax.DefineCommand("build", ref command, "builds a folder containing docfx.yml");
+                syntax.DefineOption("o|out", ref options.Output, "output folder");
+                syntax.DefineOption("log", ref options.Log, "path to log file");
+                syntax.DefineOption("locale", ref options.BuildLocale, "locale to build");
+                syntax.DefineOptionList("github-token", ref options.GitHubTokens, "GitHub personal access tokens to call github API");
```
Comment (yufeih) [inline src/docfx/cli/Program.cs]: Add `github-token`, `locale` later when we implement these features.

File: src/docfx/cli/Program.cs [hunk 2]
```diff
@@ -9,23 +10,43 @@ internal class Program
...
                     break;
                 case "build":
-                    await Build.Run(docset, options, log);
+                    await Build.Run(source, options, log);
                     break;
             }
         }
 
-        private static (string command, string docset, CommandLineOptions options) ParseCommandLineOptions(string[] args)
+        private static (string command, string source, CommandLineOptions options) ParseCommandLineOptions(string[] args)
```
Comment (OsmondJiang) [inline src/docfx/cli/Program.cs]: >source [](start = 47, length = 6)

docset would be better?

File: src/docfx/cli/Program.cs [hunk 3]
```diff
@@ -9,23 +10,43 @@ internal class Program
...
+                syntax.DefineParameter("source", ref source, "docset path that contains docfx.yml");
+
+                // Build command
+                syntax.DefineCommand("build", ref command, "builds a folder containing docfx.yml");
+                syntax.DefineOption("o|out", ref options.Output, "output folder");
+                syntax.DefineOption("log", ref options.Log, "path to log file");
+                syntax.DefineOption("locale", ref options.BuildLocale, "locale to build");
+                syntax.DefineOptionList("github-token", ref options.GitHubTokens, "GitHub personal access tokens to call github API");
+                syntax.DefineOption("stable", ref options.Stable, "produces stable output for comparison in a diff tool");
+                syntax.DefineParameter("source", ref source, "docset path that contains docfx.yml");
```
Comment (OsmondJiang) [inline src/docfx/cli/Program.cs]: >source [](start = 40, length = 6)

why it's parsed twice?

File: src/docfx/cli/Program.cs [hunk 4]
```diff
@@ -9,23 +10,43 @@ internal class Program
...
         {
-            return (args[0], args[1], default);
+            var command = string.Empty;
+            var source = ".";
+            var options = new CommandLineOptions();
+
+            ArgumentSyntax.Parse(args, syntax =>
+            {
+                // Restore command
+                syntax.DefineCommand("restore", ref command, "restores dependencies before build");
```
Comment (OsmondJiang) [inline src/docfx/cli/Program.cs]: >DefineCommand [](start = 23, length = 13)

would you please help to add a example about how to use this command to the comments?

File: src/docfx/cli/Program.cs [hunk 5]
```diff
@@ -9,23 +10,43 @@ internal class Program
...
+                    await Build.Run(source, options, log);
                     break;
             }
         }
 
-        private static (string command, string docset, CommandLineOptions options) ParseCommandLineOptions(string[] args)
+        private static (string command, string source, CommandLineOptions options) ParseCommandLineOptions(string[] args)
         {
-            return (args[0], args[1], default);
+            var command = string.Empty;
```
Comment (OsmondJiang) [inline src/docfx/cli/Program.cs]: >command [](start = 16, length = 7)

better to add a default command like "build"

File: src/docfx/cli/Program.cs (hunk 3 above)
Comment (928PJY) [reply src/docfx/cli/Program.cs]: different command have to define their own option

---
In reply to: [184253393](https://github.com/dotnet/docfx/pull/2690#discussion_r184253393) [](ancestors = 184253393)

Comment (OsmondJiang) [reply src/docfx/cli/Program.cs]: it should be a common options

---
In reply to: [184261151](https://github.com/dotnet/docfx/pull/2690#discussion_r184261151) [](ancestors = 184261151,184253393)

File: src/docfx/cli/Program.cs [hunk 6]
```diff
@@ -25,7 +26,27 @@ internal class Program
...
+                // Restore command
+                // usage: docfx restore [docset]
+                syntax.DefineCommand("restore", ref command, "restores dependencies before build");
+                syntax.DefineParameter("docset", ref docset, "docset path that contains docfx.yml");
+
+                // Build command
+                // usage: docfx build [dopcset] [-o/--out output] [--log log] [--stable]
+                syntax.DefineCommand("build", ref command, "builds a folder containing docfx.yml");
+                syntax.DefineOption("o|out", ref options.Output, "output folder");
+                syntax.DefineOption("log", ref options.Log, "path to log file");
```
Comment (superyyrrzz) [inline src/docfx/cli/Program.cs]: >log [](start = 37, length = 3)

"l|log"?

**Pull request merged**